cmd/spx/template/game.js

```diff
@@ -28,10 +31,132 @@ class GameApp {
         };
     }
 
-    Start() {
+    StartProject() {
         this.installProject();
     }
 
+    async UpdateProject(newData, addInfos, deleteInfos, updateInfos) {
+        console.log("UpdateProject", this.oldData, newData)
+        if (addInfos == undefined) {
+            const diffInfos = await this.getZipDiffInfos(this.oldData, newData);
+            addInfos = diffInfos.addInfos;
+            deleteInfos = diffInfos.deleteInfos;
+            updateInfos = diffInfos.updateInfos;
+        }
+        this.oldData = newData
+        console.log('DiffInfos :', addInfos, deleteInfos, updateInfos);
+        let mergedArray = addInfos.concat(updateInfos);
+        await this.addOrUpdateFiles(mergedArray, newData);
+        this.deleteFiles(deleteInfos);
+    }
+
+    StopGame() {
+        if(this.game != null){
+            this.game.requestQuit()
+        }
+    }
+
+    async addOrUpdateFiles(paths, zipData) {
+        const zip = new JSZip();
+        const zipContent = await zip.loadAsync(zipData);
+        let datas = []
+        for (let path of paths) {
+            const dstFile = zipContent.files[path];
+            let data = await dstFile.async('arraybuffer');
+            if (!dstFile.dir) {
+                datas.push({ "path": path, "data": data })
+            }
+        }
+        const evt = new CustomEvent('add_files', {
+            detail: datas
+        });
+        this.gameCanvas.dispatchEvent(evt);
+    }
+
+    deleteFiles(paths) {
+        paths = paths.map(info => `res://${info}`);
+        const evt = new CustomEvent('delete_files', {
+            detail: paths
+        });
+        this.gameCanvas.dispatchEvent(evt);
+    }
+
+    async getZipDiffInfos(srcZip, dstZip) {
```

Comment:
As noted in https://github.com/goplus/spx/pull/375#discussion_r1831959230, it is the Builder's responsibility to calculate and pass the diff information, so implementing `getZipDiffInfos` here may be unnecessary. A similar logic is expected to be implemented by the Builder (the parent page).

Reply:
As discussed, Builder will consistently provide diff information when calling `updateProject`, as it can obtain that data more easily and efficiently.

Reply:
https://github.com/goplus/spx/pull/376